flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java

```diff
@@ -0,0 +1,189 @@
+ * verify that programs in BATCH execution mode can be executed in stages.
+ */
+public class DataStreamBatchExecutionITCase {
+	private static final int DEFAULT_PARALLELISM = 1;
+
+	@ClassRule
+	public static MiniClusterWithClientResource miniClusterResource = new MiniClusterWithClientResource(
+			new MiniClusterResourceConfiguration.Builder()
+					.setNumberTaskManagers(1)
+					.setNumberSlotsPerTaskManager(DEFAULT_PARALLELISM)
+					.build());
+
+	/**
+	 * We induce a failure in the last mapper. In BATCH execution mode the part of the pipeline
+	 * before the key-by should not be re-executed. Only the part after that will restart. We check
+	 * that by suffixing the attempt number to records and asserting the correct number.
+	 */
+	@Test
+	public void batchFailoverWithKeyByBarrier() throws Exception {
+
+		final StreamExecutionEnvironment env = getExecutionEnvironment();
+
+		DataStreamSource<String> source = env.fromElements("foo", "bar");
+
+		SingleOutputStreamOperator<String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.keyBy(in -> in)
+				.map(new SuffixAttemptId("c"))
+				.map(new OnceFailingMapper("d"));
+
+		try (CloseableIterator<String> result = mapped.executeAndCollect()) {
+
+			// only the operators after the key-by "barrier" are restarted and will have the
+			// "attempt 1" suffix
+			assertThat(
+					iteratorToList(result),
+					containsInAnyOrder("foo-a0-b0-c1-d1", "bar-a0-b0-c1-d1"));
+		}
+	}
+
+	@Ignore
+	@Test
+	public void stagedExecutionWithDifferentSlotGroups() throws Exception {
+
+		final StreamExecutionEnvironment env = getExecutionEnvironment();
+
+		DataStreamSource<String> source = env.fromElements("ciao", "ciao");
+
+		SingleOutputStreamOperator<String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.map(new SuffixAttemptId("c"))
+				.slotSharingGroup("foo")
+				.map(new SuffixAttemptId("d"))
+				.slotSharingGroup("bar");
+
+		List<String> result = mapped.executeAndCollect(2);
+
+		// only the operators after the key-by "barrier" are restarted and will have the "attempt 1"
+		// suffix
+		assertThat(result, contains("ciao-a0-b0-c0-d0", "ciao-a0-b0-c0-d0"));
+	}
+
+	@Test
+	public void dataSetBatchFailoverWithKeyByBarrier() throws Exception {
+
+		final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+		env.setRestartStrategy(RestartStrategies.fixedDelayRestart(10, Time.milliseconds(1)));
+
+		DataSet<String> source = env.fromElements("foo", "bar");
+
+		MapOperator<String, String> mapped = source
+				.map(new SuffixAttemptId("a"))
+				.map(new SuffixAttemptId("b"))
+				.groupBy(in -> in)
+				.first(1)
+				.map(new SuffixAttemptId("c"))
+				.map(new OnceFailingMapper("d"));
+
+		List<String> result = mapped.collect();
+
+		// it seems the DataSet API restarts the whole graph here?
+		assertThat(result, containsInAnyOrder("foo-a1-b1-c1-d1", "bar-a1-b1-c1-d1"));
+	}
```

Comment:
I am late to the review, but should we leave this test, or not? The reason I am asking is because it does not verify anything related to `DataStreamBatchExecutionITCase` :)

Reply:
I just added this as an example to show how DataSet works and would not merge it.

Reply:
I see. Thanks for the explanation!